File: TP-U4/ej1.py
#Representa a cada paciente en la lista.
class Paciente:
    def __init__(self, id, nombre, edad):
        self.id = id
        self.nombre = nombre
        self.edad = edad
        self.siguiente = None

#Puntero 'head' para referenciar el primer nodo de la lista.
class ListaPacientes:
    def __init__(self):
        self.head = None

    def registrar_paciente(self, id, nombre, edad):
        nuevo_paciente = Paciente(id, nombre, edad)
        if self.head is None:
            self.head = nuevo_paciente
        else:
            actual = self.head
            while actual.siguiente:
                actual = actual.siguiente
            actual.siguiente = nuevo_paciente

    def modificar_paciente(self, id, nuevo_nombre, nueva_edad):
        actual = self.head
        while actual:
            if actual.id == id:
                actual.nombre = nuevo_nombre
                actual.edad = nueva_edad
                return
            actual = actual.siguiente
        print("Paciente no encontrado.")

    def eliminar_paciente(self, id):
        actual = self.head
        previo = None
        while actual:
            if actual.id == id:
                if previo:
                    previo.siguiente = actual.siguiente
                else:
                    self.head = actual.siguiente
                return
            previo = actual
            actual = actual.siguiente
        print("Paciente no encontrado.")

    def listar_pacientes(self):
        actual = self.head
        if not actual:
            print("No hay pacientes registrados.")
        while actual:
            print(f"ID: {actual.id}, Nombre: {actual.nombre}, Edad: {actual.edad}")
            actual = actual.siguiente

def menu():
    lista = ListaPacientes()
    while True:
        print("\n--- Menú del Sistema de Pacientes ---")
        print("1. Registrar Paciente")
        print("2. Modificar Paciente")
        print("3. Eliminar Paciente")
        print("4. Listar Pacientes")
        print("5. Salir")
        
        opcion = input("Seleccione una opción: ")
        
        if opcion == '1':
            id = int(input("Ingrese el ID del paciente: "))
            nombre = input("Ingrese el nombre del paciente: ")
            edad = int(input("Ingrese la edad del paciente: "))
            lista.registrar_paciente(id, nombre, edad)
            print("Paciente registrado exitosamente.")
        elif opcion == '2':
            id = int(input("Ingrese el ID del paciente a modificar: "))
            nuevo_nombre = input("Ingrese el nuevo nombre del paciente: ")
            nueva_edad = int(input("Ingrese la nueva edad del paciente: "))
            lista.modificar_paciente(id, nuevo_nombre, nueva_edad)
        elif opcion == '3':
            id = int(input("Ingrese el ID del paciente a eliminar: "))
            lista.eliminar_paciente(id)
        elif opcion == '4':
            lista.listar_pacientes()
        elif opcion == '5':
            print("Saliendo del sistema...")
            break
        else:
            print("Opción no válida. Por favor, intente nuevamente.")

# Ejecución del menú
menu()

File: TP-U4/ej1.cpp
#include <iostream>
#include <string>

using namespace std;

//Representa un nodo en la lista vinculada
class Paciente {
public:
    int id;
    string nombre;
    int edad;
    Paciente* siguiente;

    Paciente(int id, string nombre, int edad) 
        : id(id), nombre(nombre), edad(edad), siguiente(nullptr) {}
};

class ListaPacientes {
private:
    Paciente* head; //'head' apunta al primer nodo de la lista

public:
    ListaPacientes() : head(nullptr) {}

    ~ListaPacientes() {
        while (head) {
            Paciente* temp = head;
            head = head->siguiente;
            delete temp;
        }
    }

    void registrarPaciente(int id, string nombre, int edad) {
        Paciente* nuevoPaciente = new Paciente(id, nombre, edad);
        if (head == nullptr) {
            head = nuevoPaciente;
        } else {
            Paciente* actual = head;
            while (actual->siguiente) {
                actual = actual->siguiente;
            }
            actual->siguiente = nuevoPaciente;
        }
    }

    void modificarPaciente(int id, string nuevoNombre, int nuevaEdad) {
        Paciente* actual = head;
        while (actual) {
            if (actual->id == id) {
                actual->nombre = nuevoNombre;
                actual->edad = nuevaEdad;
                return;
            }
            actual = actual->siguiente;
        }
        cout << "Paciente no encontrado.\n";
    }

    void eliminarPaciente(int id) {
        Paciente* actual = head;
        Paciente* previo = nullptr;
        while (actual) {
            if (actual->id == id) {
                if (previo) {
                    previo->siguiente = actual->siguiente;
                } else {
                    head = actual->siguiente;
                }
                delete actual;
                return;
            }
            previo = actual;
            actual = actual->siguiente;
        }
        cout << "Paciente no encontrado.\n";
    }

    void listarPacientes() const {
        if (head == nullptr) {
            cout << "No hay pacientes registrados.\n";
            return;
        }
        Paciente* actual = head;
        while (actual) {
            cout << "ID: " << actual->id 
                      << ", Nombre: " << actual->nombre 
                      << ", Edad: " << actual->edad << '\n';
            actual = actual->siguiente;
        }
    }
};

void menu() {
    ListaPacientes lista;
    while (true) {
        cout << "\n--- Menú del Sistema de Pacientes ---\n";
        cout << "1. Registrar Paciente\n";
        cout << "2. Modificar Paciente\n";
        cout << "3. Eliminar Paciente\n";
        cout << "4. Listar Pacientes\n";
        cout << "5. Salir\n";
        
        int opcion;
        cout << "Seleccione una opción: ";
        cin >> opcion;
        cin.ignore(); // Limpiar el buffer de entrada
        
        switch (opcion) {
            case 1: {
                int id;
                string nombre;
                int edad;
                cout << "Ingrese el ID del paciente: ";
                cin >> id;
                cin.ignore();
                cout << "Ingrese el nombre del paciente: ";
                getline(cin, nombre);
                cout << "Ingrese la edad del paciente: ";
                cin >> edad;
                lista.registrarPaciente(id, nombre, edad);
                cout << "Paciente registrado exitosamente.\n";
                break;
            }
            case 2: {
                int id;
                string nuevoNombre;
                int nuevaEdad;
                cout << "Ingrese el ID del paciente a modificar: ";
                cin >> id;
                cin.ignore();
                cout << "Ingrese el nuevo nombre del paciente: ";
                getline(cin, nuevoNombre);
                cout << "Ingrese la nueva edad del paciente: ";
                cin >> nuevaEdad;
                lista.modificarPaciente(id, nuevoNombre, nuevaEdad);
                break;
            }
            case 3: {
                int id;
                cout << "Ingrese el ID del paciente a eliminar: ";
                cin >> id;
                lista.eliminarPaciente(id);
                break;
            }
            case 4:
                lista.listarPacientes();
                break;
            case 5:
                cout << "Saliendo del sistema...\n";
                return;
            default:
                cout << "Opción no válida. Por favor, intente nuevamente.\n";
                break;
        }
    }
}

int main() {
    menu();
    return 0;
}
